torch/distributed/distributed_c10d.py

```diff
@@ -1465,6 +1466,62 @@ def reduce_scatter(output,
         work.wait()
 
 
+def alltoall(output,
+             input,
+             group=group.WORLD,
+             async_op=False):
+    """
+    Each process splits input tensor equally and then scatters the split list to all processes in a group.
```

Comment:
what if the tensor cannot be equally split across world_size? e.g., input is [0, 1, 2, 3], and world_size = 3? Will it error out before running the allToAll op, or does different ranks need to provide different size output tensors?

Reply:
At present, it expects to split both input and output exactly but code doesn't check for it. However, ideally we want to support the case where input and/or output does not split equally. 
I propose adding `input_split_sizes` and `output_split_sizes` optional arguments to provide list of per rank integer split sizes when they are not equally split (something similar to what `torch.split()` uses). Another question is, if input/output tensors are multi-dimensional, do we split whole tensor or just `dim 0` and what the split sizes would map to? Again, my recommendation is to assume `dim 0` implicitly as other dims would cause non-contiguous splits and require copies anyway. So, better we use explicit list of tensors when split is not along `dim 0`. In either case, sum of split sizes should be less than or equal to total tensor size or `dim 0` size depending one which behavior we pick.
Finally, I don't know if there is use case for it or not, but if we want to match [MPI_Alltoallv](https://www.open-mpi.org/doc/v4.0/man3/MPI_Alltoallv.3.php) semantic, we may also add optional list of offsets for input and output tensors from where each split starts.

Let me know your thoughts and I will update PR accordingly.

PS: Similar discussion applies to all_gather_base, scatter_base and gather_base.

Reply:
I have one more question regarding APIs:

```
  virtual std::shared_ptr<ProcessGroup::Work> allgather(
      std::vector<std::vector<at::Tensor>>& outputTensors,
      std::vector<at::Tensor>& inputTensors,
      const AllgatherOptions& opts = AllgatherOptions()) = 0;

  virtual std::shared_ptr<ProcessGroup::Work> allgather_base(
      at::Tensor& outputBuffer,
      at::Tensor& inputBuffer,
      const AllgatherOptions& opts = AllgatherOptions()) = 0;
```

These are the abstract definitions of `ProcessGroup::allgather` and `ProcessGroup::allgather_base`. I see before adding `allgather_base`, each `ProcessGroup` method was taking either `vector<Tensor>` or `vector<vector<Tensor>>` as arguments. But all gather simply take single tensor as argument. Is this part of planned API clean up or `allgather_base` is expected to take a `vector<Tensor>` as argument?

Reply:
Hey @agolynski, could you please comment on the plans for `allgather` and `allgather_base`?

Reply:
Hi @mrshenli @agolynski, 
I have updated PR with my proposal to support uneven split. Please take a look. I have also added more examples to documentation of API.
cc @dmudiger @srinivas212

Reply:
Looks good.